github-stats/fetcher/client.go

```diff
@@ -0,0 +1,257 @@
+		return client.githubClient, nil
+	}
+	token := client.Token
+	if len(token) == 0 && len(client.TokenFile) != 0 {
+		data, err := ioutil.ReadFile(client.TokenFile)
+		if err != nil {
+			return nil, err
+		}
+		token = strings.TrimSpace(string(data))
+	}
+
+	if len(token) > 0 {
+		ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token})
+		tc := oauth2.NewClient(oauth2.NoContext, ts)
+		client.githubClient = github.NewClient(tc)
+	} else {
+		client.githubClient = github.NewClient(nil)
+	}
+	return client.githubClient, nil
+}
+
+// Make sure we have not reached the limit or wait
+func (client *Client) limitsCheckAndWait() {
+	var sleep time.Duration
+	githubClient, err := client.getGithubClient()
+	if err != nil {
+		glog.Error("Failed to get RateLimits: ", err)
+		sleep = time.Minute
+	} else {
+		limits, _, err := githubClient.RateLimits()
+		if err != nil {
+			glog.Error("Failed to get RateLimits:", err)
+			sleep = time.Minute
+		}
+		if limits != nil && limits.Core != nil && limits.Core.Remaining < tokenLimit {
+			sleep = limits.Core.Reset.Sub(time.Now())
+			glog.Warning("RateLimits: reached. Sleeping for ", sleep)
+		}
+	}
+
+	time.Sleep(sleep)
+}
+
+// ClientInterface describes what a client should be able to do
+type ClientInterface interface {
+	FetchIssues(time.Time, chan *github.Issue)
+	FetchIssueEvents(*int, chan *github.IssueEvent)
+	FetchIssueComments(int, time.Time, chan *github.IssueComment)
+	FetchPullComments(int, time.Time, chan *github.PullRequestComment)
+}
+
+// FetchIssues from Github, until 'latest' time
+func (client *Client) FetchIssues(latest time.Time, c chan *github.Issue) {
+	opt := &github.IssueListByRepoOptions{Since: latest, Sort: "updated", State: "all", Direction: "asc"}
+
+	githubClient, err := client.getGithubClient()
+	if err != nil {
+		close(c)
+		glog.Error(err)
+		return
+	}
+
+	for {
+		client.limitsCheckAndWait()
+
+		issues, resp, err := githubClient.Issues.ListByRepo(client.Org, client.Project, opt)
+		if err != nil {
+			close(c)
+			glog.Error(err)
+			return
+		}
+
+		for _, issue := range issues {
+			glog.Info("Issue: ", *issue.Number, ", last updated: ", *issue.UpdatedAt)
+			c <- issue
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opt.ListOptions.Page = resp.NextPage
+	}
+
+	close(c)
+}
+
+// Look for a specific Id in a list of events
+func wasIDFound(events []*github.IssueEvent, ID int) bool {
+	for _, event := range events {
+		if *event.ID == ID {
+			return true
+		}
+	}
+	return false
+}
+
+// FetchIssueEvents from github and return the full list, until it matches 'latest'
+// The entire last page will be included so you can have redundancy.
+func (client *Client) FetchIssueEvents(latest *int, c chan *github.IssueEvent) {
+	opt := &github.ListOptions{PerPage: 100}
+
+	githubClient, err := client.getGithubClient()
+	if err != nil {
+		close(c)
+		glog.Error(err)
+		return
+	}
+
+	for {
+		client.limitsCheckAndWait()
+
+		glog.Info("Downloading events page: ", opt.Page)
```

Comment:
While the other Info was hard to read, I think it wouldn't make this one any better